test/export/StructureDefinitionExporter.test.ts

```diff
@@ -609,6 +609,120 @@ describe('StructureDefinitionExporter', () => {
     expect(barSlice).toBeDefined();
   });
 
+  it('should apply a ContainsRule of a defined extension on an extension element', () => {
+    const profile = new Profile('Foo');
+    profile.parent = 'Observation';
+
+    const rule = new ContainsRule('extension');
+    rule.items = ['valueset-expression'];
+    profile.rules.push(rule);
+
+    exporter.exportStructDef(profile);
+    const sd = exporter.structDefs[0];
+
+    const extension = sd.elements.find(e => e.id === 'Observation.extension');
+    const valuesetExpression = sd.elements.find(
+      e => e.id === 'Observation.extension:valueset-expression'
+    );
+
+    expect(extension.slicing).toBeDefined();
+    expect(extension.slicing.discriminator.length).toBe(1);
+    expect(extension.slicing.discriminator[0]).toEqual({ type: 'value', path: 'url' });
+    expect(valuesetExpression).toBeDefined();
+    expect(valuesetExpression.type[0]).toEqual({
+      code: 'Extension',
+      profile: ['http://hl7.org/fhir/StructureDefinition/valueset-expression']
+    });
+  });
+
+  it('should apply a ContainsRule of a defined extension on a modifierExtension element', () => {
+    const profile = new Profile('Foo');
+    profile.parent = 'Observation';
+
+    const rule = new ContainsRule('modifierExtension');
+    rule.items = ['valueset-expression'];
+    profile.rules.push(rule);
+
+    exporter.exportStructDef(profile);
+    const sd = exporter.structDefs[0];
+
+    const extension = sd.elements.find(e => e.id === 'Observation.modifierExtension');
+    const valuesetExpression = sd.elements.find(
+      e => e.id === 'Observation.modifierExtension:valueset-expression'
+    );
+
+    expect(extension.slicing).toBeDefined();
+    expect(extension.slicing.discriminator.length).toBe(1);
+    expect(extension.slicing.discriminator[0]).toEqual({ type: 'value', path: 'url' });
+    expect(valuesetExpression).toBeDefined();
+    expect(valuesetExpression.type[0]).toEqual({
+      code: 'Extension',
+      profile: ['http://hl7.org/fhir/StructureDefinition/valueset-expression']
+    });
+  });
+
+  it('should apply a ContainsRule of an undefined extension on an extension element', () => {
+    const profile = new Profile('Foo');
+    profile.parent = 'Observation';
+
+    const rule = new ContainsRule('extension');
+    rule.items = ['foo'];
+    profile.rules.push(rule);
```

Comment:
Technically, this is an invalid SD since the `slicing` says to slice on `url`, but the slice does not set `extension[foo] ^type[0].profile` or fix `extension[foo].url`.  It seems like maybe we should actually detect this situation and report an error?  Otherwise it seems easy for the author to fat-finger what is meant to be an existing extension name/id, but not realize it since we happily accept it and treat it like a new (undefined) extension.

Reply:
Hmm. This is a good point, but I am not sure how to go about it. Since we don't know when that rule that sets the url will be. It'll be sometime after the contains rule, but not necessarily immediately after.